File: app/src/main/java/se/gustavkarlsson/aurora_notifier/android/dagger/modules/WeatherModule.java
package se.gustavkarlsson.aurora_notifier.android.dagger.modules;

import android.content.Context;

import dagger.Module;
import dagger.Provides;
import dagger.Reusable;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;
import se.gustavkarlsson.aurora_notifier.android.R;
import se.gustavkarlsson.aurora_notifier.android.background.providers.WeatherProvider;
import se.gustavkarlsson.aurora_notifier.android.background.providers.impl.OpenWeatherMapService;
import se.gustavkarlsson.aurora_notifier.android.background.providers.impl.RetrofittedOpenWeatherMapProvider;

@Module
public abstract class WeatherModule {
	private static final String API_URL = "http://api.openweathermap.org/data/2.5/";

	@Provides
	@Reusable
	static OpenWeatherMapService provideOpenWeatherMapService() {
		return new Retrofit.Builder()
				.baseUrl(API_URL)
				.addConverterFactory(GsonConverterFactory.create())
				.build().create(OpenWeatherMapService.class);
	}

	@Provides
	@Reusable
	static WeatherProvider provideWeatherProvider(Context context, OpenWeatherMapService openWeatherMapService) {
		String apiKey = context.getString(R.string.api_key_openweathermap);
		return new RetrofittedOpenWeatherMapProvider(openWeatherMapService, apiKey);
	}
}

File: app/src/main/kotlin/se/gustavkarlsson/skylight/android/WeatherModule.kt
package se.gustavkarlsson.skylight.android

import io.reactivex.Flowable
import io.reactivex.schedulers.Schedulers
import org.koin.dsl.module.module
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import se.gustavkarlsson.skylight.android.entities.Location
import se.gustavkarlsson.skylight.android.entities.Weather
import se.gustavkarlsson.skylight.android.extensions.create
import se.gustavkarlsson.skylight.android.extensions.minutes
import se.gustavkarlsson.skylight.android.extensions.seconds
import se.gustavkarlsson.skylight.android.services.Streamable
import se.gustavkarlsson.skylight.android.services.providers.WeatherProvider
import se.gustavkarlsson.skylight.android.services_impl.providers.RetrofittedOpenWeatherMapWeatherProvider
import se.gustavkarlsson.skylight.android.services_impl.providers.openweathermap.OpenWeatherMapApi
import se.gustavkarlsson.skylight.android.services_impl.streamables.WeatherProviderStreamable

val weatherModule = module {

	single<OpenWeatherMapApi> {
		Retrofit.Builder()
			.baseUrl("http://api.openweathermap.org/data/2.5/")
			.addConverterFactory(GsonConverterFactory.create())
			.addCallAdapterFactory(RxJava2CallAdapterFactory.createWithScheduler(Schedulers.io()))
			.build().create()
	}

	single<WeatherProvider> {
		RetrofittedOpenWeatherMapWeatherProvider(get(), BuildConfig.OPENWEATHERMAP_API_KEY, 5)
	}

	single<Streamable<Weather>>("weather") {
		val locations = get<Flowable<Location>>("location")
		WeatherProviderStreamable(locations, get(), 15.minutes, 10.seconds)
	}

	single<Flowable<Weather>>("weather") {
		get<Streamable<Weather>>("weather")
			.stream
			.replay(1)
			.refCount()
	}

}
